File: leet_easy_c++/leet_141/leet_141b.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        unordered_map<ListNode*, int> umap;
        ListNode * fast = head;
        ListNode * slow = head;
        while(fast){
            if(fast->next){
                if(fast->next == slow || fast->next->next == slow) return true;
                fast = fast->next->next;
                slow = slow->next;
            }
            else return false;
        }
        return false;
    }
};

File: leet_easy_python/leet_141/leet_141b.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        dic = dict()
        while head:
            if head in dic: return True
            dic[head] = 1
            head = head.next
        return False
        
        """
        :type head: ListNode
        :rtype: bool
        """
        